histogram/src/lib.rs

```diff
@@ -1,206 +1,121 @@
+        let mut percentiles = percentiles.to_vec();
+        percentiles.sort_by(|a, b| a.partial_cmp(b).unwrap());
+
+        // get a sorted list of the counts needed for each percentile
+        let counts: Vec<u128> = percentiles
             .iter()
-            .map(|p| (p.bucket().low(), p.bucket().high()))
+            .map(|p| (p / 100.0 * total as f64).ceil() as u128)
             .collect();
 
-        // this histogram config has enough resolution to keep the error lower
-        assert_eq!(
-            &percentiles,
-            &[
-                (11, 11),
-                (100, 103),
-                (256, 271),
-                (512, 543),
-                (768, 799),
-                (896, 927),
-                (992, 1023)
-            ]
-        );
-    }
-
-    #[test]
-    fn percentiles_3() {
-        let histogram = Histogram::builder().build().unwrap();
-        histogram.increment(1, 1).unwrap();
-        histogram.increment(10000000, 1).unwrap();
-
-        let percentiles = histogram.percentiles(&[25.0, 75.0]).unwrap();
-
-        assert_eq!(histogram.percentile(25.0).map(|b| b.high()), Ok(1));
-        assert_eq!(histogram.percentile(75.0).map(|b| b.high()), Ok(10010623));
-
-        assert_eq!(percentiles.get(0).map(|b| b.bucket().high()), Some(1));
-        assert_eq!(
-            percentiles.get(1).map(|b| b.bucket().high()),
-            Some(10010623)
-        );
-    }
-
-    #[test]
-    fn increment_and_decrement() {
-        let histogram = Histogram::builder().build().unwrap();
-        assert_eq!(
-            histogram.percentile(0.0).map(|b| b.count()),
-            Err(Error::Empty)
-        );
-
-        histogram.increment(1, 1).unwrap();
-        assert_eq!(histogram.percentile(0.0).map(|b| b.count()), Ok(1));
-
-        histogram.decrement(1, 1).unwrap();
-        assert_eq!(
-            histogram.percentile(0.0).map(|b| b.count()),
-            Err(Error::Empty)
-        );
-    }
+        // a vec to hold the results
+        let mut result = Vec::with_capacity(percentiles.len());
+
+        let mut percentile_idx = 0;
+        let mut bucket_idx = 0;
+        let mut partial_sum = self.get_count(bucket_idx) as u128;
+
+        while percentile_idx < percentiles.len() {
+            let percentile = percentiles[percentile_idx];
+            let count = counts[percentile_idx];
+
+            // repeatedly push percentile-bucket pairs into the result while our
+            // current partial sum fulfills the count needed for each percentile
+            if partial_sum >= count {
+                result.push((percentile, self.get_bucket(bucket_idx).unwrap()));
+                percentile_idx += 1;
+                continue;
+            }
+
+            // increment the partial sum by the count of the next bucket
+            bucket_idx += 1;
+            if bucket_idx >= self.config().total_bins() {
+                break;
```

Comment:
I think we only hit this case if a percentile > 100 is requested -- should we also filter those out while converting percentiles to counts?

Reply:
We should return an error if percentile as f64 > 100.0 - in fact there is an error type defined for this, though I think the logic that enforces the constraints is missing...

Reply:
Fixed.